releasing scrollViewUpdater on pan .ended, to stop observing

soreshko opened the pull request:
This prevents crashing if DeckPresentationController contains a table view, since scrollViewUpdater observes contentOffset and results in crash when table view displayed is eventually released while still having this KVO observer:

Fatal Exception: NSInternalInconsistencyException
An instance 0x132094a00 of class UITableView was deallocated while key value observers were still registered with it. Current observation info: <NSKeyValueObservationInfo 0x174626680> ( <NSKeyValueObservance 0x17485da60: Observer: 0x175e74cc0, Key path: contentOffset, Options: <New: NO, Old: NO, Prior: NO> Context: 0x0, Property: 0x170a50890> )

HarshilShah commented:
Good catch, and thank you for the PR!

soreshko commented:
Thank you for creating this great library!

Best regards,
Sergey

On March 8, 2018 at 10:39:32 PM, Harshil Shah (notifications@github.com)
wrote:

Good catch, and thank you for the PR!

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/HarshilShah/DeckTransition/pull/66#issuecomment-371703913>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AGE07NA-Ep7w93dO5bts_kfpjugyMYXLks5tcflxgaJpZM4Sjd7V>
.

HarshilShah commented:
Hi, sorry this is late, but could you possibly upload a demo project that reproduces this? I've just now found an edge case that breaks with this fix, and can't reproduce the bug myself.

soreshko commented:
Hi Harshil,

The demo project may take some effort, as we used it in a large project
with many dependencies (e.g. tableView in question was managed by TableKit
library, although theoretically this should not have affected the bug).

In the mean time, if you can point me to the code/branch with the
simplified version, I can test in our app if the change eliminates the
error, thus making my fix irrelevant.

Best regards,
Sergey

On March 25, 2018 at 6:23:02 AM, Harshil Shah (notifications@github.com)
wrote:

Getting rid of a bunch of the stuff that isn't required here, this is a
simplified version of the ScrollViewUpdater code which I just ran in a
Playground. Seems to work with no error.

class ScrollViewUpdater {

    var observation: NSKeyValueObservation?

    init(scrollView: UIScrollView) {
        self.observation = scrollView.observe(\.contentOffset,
options: [.initial], changeHandler: { [weak self] _, _ in
            print("Testing")
        })
    }

    deinit {
        observation = nil
    }
}
class ViewController: UIViewController {
    var tableView = UITableView()
}
var controller: ViewController? = ViewController()let updater =
ScrollViewUpdater(scrollView: controller!.tableView)
controller = nil

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/HarshilShah/DeckTransition/pull/66#issuecomment-375959730>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AGE07M-F6_zzbq_sAgfWaSBGj2uFLiDhks5th3ACgaJpZM4Sjd7V>
.

HarshilShah commented:
There’s no simplified version, it’s the previous version before this PR that doesn’t crash in my testing. The code above was just to show that I can’t reproduce the crash by releasing a UITableView.

It’d be helpful if you could reproduce it because while this PR does appear to fix it, it also breaks something else and I’d much rather see if there’s another solution, for which I’d need to be able to reproduce the problem first.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

soreshko commented:
OK, got it, I’ll work on this today and tomorrow.

S.


On March 26, 2018 at 7:05:55 AM, Harshil Shah (notifications@github.com)
wrote:

There’s no simplified version, it’s the previous version before this PR
that doesn’t crash in my testing. The code above was just to show that I
can’t reproduce the crash by releasing a UITableView.

It’d be helpful if you could reproduce it because while this PR does appear
to fix it, it also breaks something else and I’d much rather see if there’s
another solution, for which I’d need to be able to reproduce the problem
first.

Sent with GitHawk <http://githawk.com>

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/HarshilShah/DeckTransition/pull/66#issuecomment-376128935>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AGE07HKuo8a29fQrYHN0Dfb85Ra-md00ks5tiMuPgaJpZM4Sjd7V>
.
